src/compiler/builder.ts

```diff
@@ -77,8 +77,8 @@ namespace ts {
          */
         onUpdateSourceFile(program: Program, sourceFile: SourceFile): void;
         /**
-         * Called when source file has not changed but has some of the resolutions invalidated

```

Comment:
"Called when source file has not changed" sounds strange -- it's unusual to invoke a callback when something doesn't happen. How about:

> For all source files, either "onUpdateSourceFile" or "onUpdateSourceFileWithSameVersion" will be called.
> If the builder is sure that the source file needs an update, "onUpdateSourceFile" will be called;
> otherwise "onUpdateSourceFileWithSameVersion" will be called.
> This should return whether the source file should be marked as changed (meaning that something associated with file has changed, e.g. module resolution).